File: benchmarks/go/bench_branch_mix.go
package main

import (
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"time"
)

var sinkBranchMix float64

func parseArg(args []string, idx int, def int) int {
	if idx >= len(args) {
		return def
	}
	v, err := strconv.Atoi(args[idx])
	if err != nil || v <= 0 {
		return def
	}
	return v
}

func coefficientOfVariationPct(samples []int64) float64 {
	if len(samples) < 2 {
		return 0
	}
	mean := 0.0
	for _, sample := range samples {
		mean += float64(sample)
	}
	mean /= float64(len(samples))
	if mean == 0 {
		return 0
	}
	variance := 0.0
	for _, sample := range samples {
		delta := float64(sample) - mean
		variance += delta * delta
	}
	variance /= float64(len(samples))
	return math.Sqrt(variance) * 100 / mean
}

func runOnce(n int, reps int) float64 {
	arr := make([]float64, n)
	for i := range arr {
		arr[i] = float64(i)
	}
	sum := 0.0
	var state int64
	for r := 0; r < reps; r++ {
		for _, value := range arr {
			state += 17
			if state >= 97 {
				state -= 97
			}
			if state < 48 {
				sum += value
			} else {
				sum -= value
			}
		}
	}
	sinkBranchMix = sum
	return sum
}

func main() {
	n := parseArg(os.Args, 1, 100000)
	iters := parseArg(os.Args, 2, 200)
	warmupMs := parseArg(os.Args, 3, 100)
	reps := parseArg(os.Args, 4, 50)

	warmupEnd := time.Now().Add(time.Duration(warmupMs) * time.Millisecond)
	for time.Now().Before(warmupEnd) {
		runOnce(n, reps)
	}

	samples := make([]int64, iters)
	checksum := 0.0
	for it := 0; it < iters; it++ {
		start := time.Now()
		checksum = runOnce(n, reps)
		samples[it] = time.Since(start).Nanoseconds()
	}

	cvPct := coefficientOfVariationPct(samples)
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	median := samples[iters/2]
	p95 := samples[(iters*95)/100]
	ops := 0.0
	if median > 0 {
		ops = 1e9 / float64(median)
	}
	fmt.Printf("[GO BENCH] branch_mix median=%d ns/op (%.0f ops/sec), p95=%d ns/op cv_pct=%.4f | iters=%d warmup=%dms reps=%d checksum=%.17g\n",
		median, ops, p95, cvPct, iters, warmupMs, reps, checksum)
}

File: benchmarks/rust/src/bin/bench_branch_mix.rs
use std::env;
use std::hint::black_box;
use std::time::{Duration, Instant};

fn parse_arg(args: &[String], idx: usize, default: usize) -> usize {
    args.get(idx)
        .and_then(|s| s.parse::<usize>().ok())
        .filter(|v| *v > 0)
        .unwrap_or(default)
}

fn coefficient_of_variation_pct(samples: &[u64]) -> f64 {
    if samples.len() < 2 {
        return 0.0;
    }
    let mean = samples.iter().map(|sample| *sample as f64).sum::<f64>() / samples.len() as f64;
    if mean == 0.0 {
        return 0.0;
    }
    let variance = samples
        .iter()
        .map(|sample| {
            let delta = *sample as f64 - mean;
            delta * delta
        })
        .sum::<f64>()
        / samples.len() as f64;
    variance.sqrt() * 100.0 / mean
}

fn run_once(n: usize, reps: usize) -> f64 {
    let mut arr = vec![0.0_f64; n];
    for (i, item) in arr.iter_mut().enumerate() {
        *item = i as f64;
    }
    let mut sum = 0.0_f64;
    let mut state = 0_i64;
    for _ in 0..reps {
        for value in &arr {
            state += 17;
            if state >= 97 {
                state -= 97;
            }
            if state < 48 {
                sum += *value;
            } else {
                sum -= *value;
            }
        }
    }
    black_box(sum)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n = parse_arg(&args, 1, 100_000);
    let iters = parse_arg(&args, 2, 200);
    let warmup_ms = parse_arg(&args, 3, 100);
    let reps = parse_arg(&args, 4, 50);

    let warmup_end = Instant::now() + Duration::from_millis(warmup_ms as u64);
    while Instant::now() < warmup_end {
        black_box(run_once(n, reps));
    }

    let mut samples: Vec<u64> = Vec::with_capacity(iters);
    let mut checksum = 0.0_f64;
    for _ in 0..iters {
        let start = Instant::now();
        checksum = run_once(n, reps);
        samples.push(start.elapsed().as_nanos() as u64);
    }

    let cv_pct = coefficient_of_variation_pct(&samples);
    samples.sort_unstable();
    let median = samples[iters / 2];
    let p95 = samples[(iters * 95) / 100];
    let ops = if median > 0 {
        1e9_f64 / median as f64
    } else {
        0.0_f64
    };

    println!(
        "[RUST BENCH] branch_mix median={} ns/op ({:.0} ops/sec), p95={} ns/op cv_pct={:.4} | iters={} warmup={}ms reps={} checksum={:.17}",
        median, ops, p95, cv_pct, iters, warmup_ms, reps, checksum
    );
}
